{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data process and cleaning for VIPR database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "from IPython.display import FileLink\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata_path = \"[YOUR PATH TO THE METADATA]\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "data = pd.read_csv(metadata_path, sep=\"\\t\", low_memory=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Genome ID</th>\n",
       "      <th>Genome Name</th>\n",
       "      <th>Other Names</th>\n",
       "      <th>NCBI Taxon ID</th>\n",
       "      <th>Taxon Lineage IDs</th>\n",
       "      <th>Taxon Lineage Names</th>\n",
       "      <th>Superkingdom</th>\n",
       "      <th>Kingdom</th>\n",
       "      <th>Phylum</th>\n",
       "      <th>Class</th>\n",
       "      <th>...</th>\n",
       "      <th>Host Age</th>\n",
       "      <th>Host Health</th>\n",
       "      <th>Host Group</th>\n",
       "      <th>Lab Host</th>\n",
       "      <th>Passage</th>\n",
       "      <th>Other Clinical</th>\n",
       "      <th>Additional Metadata</th>\n",
       "      <th>Comments</th>\n",
       "      <th>Date Inserted</th>\n",
       "      <th>Date Modified</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10407.119236</td>\n",
       "      <td>Hepatitis B virus Viet_ocHBV111</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10407</td>\n",
       "      <td>10239;2559587;2732397;2732409;2732514;2732515;...</td>\n",
       "      <td>Viruses;Riboviria;Pararnavirae;Artverviricota;...</td>\n",
       "      <td>Viruses</td>\n",
       "      <td>Pararnavirae</td>\n",
       "      <td>Artverviricota</td>\n",
       "      <td>Revtraviricetes</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2022-02-17T04:33:34.549Z</td>\n",
       "      <td>2022-02-17T04:33:34.549Z</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10407.119237</td>\n",
       "      <td>Hepatitis B virus Viet_ocHBV68</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10407</td>\n",
       "      <td>10239;2559587;2732397;2732409;2732514;2732515;...</td>\n",
       "      <td>Viruses;Riboviria;Pararnavirae;Artverviricota;...</td>\n",
       "      <td>Viruses</td>\n",
       "      <td>Pararnavirae</td>\n",
       "      <td>Artverviricota</td>\n",
       "      <td>Revtraviricetes</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2022-02-17T04:33:34.722Z</td>\n",
       "      <td>2022-02-17T04:33:34.722Z</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>10407.119238</td>\n",
       "      <td>Hepatitis B virus Viet_ocHBV353</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10407</td>\n",
       "      <td>10239;2559587;2732397;2732409;2732514;2732515;...</td>\n",
       "      <td>Viruses;Riboviria;Pararnavirae;Artverviricota;...</td>\n",
       "      <td>Viruses</td>\n",
       "      <td>Pararnavirae</td>\n",
       "      <td>Artverviricota</td>\n",
       "      <td>Revtraviricetes</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2022-02-17T04:33:34.739Z</td>\n",
       "      <td>2022-02-17T04:33:34.739Z</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>10407.119239</td>\n",
       "      <td>Hepatitis B virus Viet_ocHBV261</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10407</td>\n",
       "      <td>10239;2559587;2732397;2732409;2732514;2732515;...</td>\n",
       "      <td>Viruses;Riboviria;Pararnavirae;Artverviricota;...</td>\n",
       "      <td>Viruses</td>\n",
       "      <td>Pararnavirae</td>\n",
       "      <td>Artverviricota</td>\n",
       "      <td>Revtraviricetes</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2022-02-17T04:33:34.911Z</td>\n",
       "      <td>2022-02-17T04:33:34.911Z</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10407.119240</td>\n",
       "      <td>Hepatitis B virus Viet_ocHBV257</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10407</td>\n",
       "      <td>10239;2559587;2732397;2732409;2732514;2732515;...</td>\n",
       "      <td>Viruses;Riboviria;Pararnavirae;Artverviricota;...</td>\n",
       "      <td>Viruses</td>\n",
       "      <td>Pararnavirae</td>\n",
       "      <td>Artverviricota</td>\n",
       "      <td>Revtraviricetes</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2022-02-17T04:36:35.044Z</td>\n",
       "      <td>2022-02-17T04:36:35.044Z</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 87 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Genome ID                      Genome Name  Other Names  NCBI Taxon ID  \\\n",
       "0  10407.119236  Hepatitis B virus Viet_ocHBV111          NaN          10407   \n",
       "1  10407.119237   Hepatitis B virus Viet_ocHBV68          NaN          10407   \n",
       "2  10407.119238  Hepatitis B virus Viet_ocHBV353          NaN          10407   \n",
       "3  10407.119239  Hepatitis B virus Viet_ocHBV261          NaN          10407   \n",
       "4  10407.119240  Hepatitis B virus Viet_ocHBV257          NaN          10407   \n",
       "\n",
       "                                   Taxon Lineage IDs  \\\n",
       "0  10239;2559587;2732397;2732409;2732514;2732515;...   \n",
       "1  10239;2559587;2732397;2732409;2732514;2732515;...   \n",
       "2  10239;2559587;2732397;2732409;2732514;2732515;...   \n",
       "3  10239;2559587;2732397;2732409;2732514;2732515;...   \n",
       "4  10239;2559587;2732397;2732409;2732514;2732515;...   \n",
       "\n",
       "                                 Taxon Lineage Names Superkingdom  \\\n",
       "0  Viruses;Riboviria;Pararnavirae;Artverviricota;...      Viruses   \n",
       "1  Viruses;Riboviria;Pararnavirae;Artverviricota;...      Viruses   \n",
       "2  Viruses;Riboviria;Pararnavirae;Artverviricota;...      Viruses   \n",
       "3  Viruses;Riboviria;Pararnavirae;Artverviricota;...      Viruses   \n",
       "4  Viruses;Riboviria;Pararnavirae;Artverviricota;...      Viruses   \n",
       "\n",
       "        Kingdom          Phylum            Class  ... Host Age Host Health  \\\n",
       "0  Pararnavirae  Artverviricota  Revtraviricetes  ...      NaN         NaN   \n",
       "1  Pararnavirae  Artverviricota  Revtraviricetes  ...      NaN         NaN   \n",
       "2  Pararnavirae  Artverviricota  Revtraviricetes  ...      NaN         NaN   \n",
       "3  Pararnavirae  Artverviricota  Revtraviricetes  ...      NaN         NaN   \n",
       "4  Pararnavirae  Artverviricota  Revtraviricetes  ...      NaN         NaN   \n",
       "\n",
       "  Host Group Lab Host Passage Other Clinical Additional Metadata  Comments  \\\n",
       "0      Human      NaN     NaN            NaN                 NaN       NaN   \n",
       "1      Human      NaN     NaN            NaN                 NaN       NaN   \n",
       "2      Human      NaN     NaN            NaN                 NaN       NaN   \n",
       "3      Human      NaN     NaN            NaN                 NaN       NaN   \n",
       "4      Human      NaN     NaN            NaN                 NaN       NaN   \n",
       "\n",
       "              Date Inserted             Date Modified  \n",
       "0  2022-02-17T04:33:34.549Z  2022-02-17T04:33:34.549Z  \n",
       "1  2022-02-17T04:33:34.722Z  2022-02-17T04:33:34.722Z  \n",
       "2  2022-02-17T04:33:34.739Z  2022-02-17T04:33:34.739Z  \n",
       "3  2022-02-17T04:33:34.911Z  2022-02-17T04:33:34.911Z  \n",
       "4  2022-02-17T04:36:35.044Z  2022-02-17T04:36:35.044Z  \n",
       "\n",
       "[5 rows x 87 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Cleaning of Initial VIPR data\n",
    "In this section we will clean the dataset to: \n",
    "- Optimize memory\n",
    "- Delete unnecessary columns\n",
    "- Rename columns to macth Genomic Dash framework template\n",
    "- Standardize African country names\n",
    "- Writing GenBank IDs to a .txt file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Genome ID', 'Genome Name', 'Other Names', 'NCBI Taxon ID',\n",
       "       'Taxon Lineage IDs', 'Taxon Lineage Names', 'Superkingdom', 'Kingdom',\n",
       "       'Phylum', 'Class', 'Order', 'Family', 'Genus', 'Species',\n",
       "       'Genome Status', 'Strain', 'Serovar', 'Biovar', 'Pathovar', 'MLST',\n",
       "       'Segment', 'Subtype', 'H_type', 'N_type', 'H1 Clade Global',\n",
       "       'H1 Clade US', 'H5 Clade', 'pH1N1-like', 'Lineage', 'Clade', 'Subclade',\n",
       "       'Other Typing', 'Culture Collection', 'Type Strain', 'Reference',\n",
       "       'Genome Quality', 'Completion Date', 'Publication', 'Authors',\n",
       "       'BioProject Accession', 'BioSample Accession', 'Assembly Accession',\n",
       "       'SRA Accession', 'GenBank Accessions', 'Sequencing Center',\n",
       "       'Sequencing Status', 'Sequencing Platform', 'Sequencing Depth',\n",
       "       'Assembly Method', 'Chromosome', 'Plasmids', 'Contigs', 'Size',\n",
       "       'GC Content', 'Contig L50', 'Contig N50', 'TRNA', 'RRNA', 'Mat Peptide',\n",
       "       'CDS', 'Coarse Consistency', 'Fine Consistency', 'CheckM Contamination',\n",
       "       'CheckM Completeness', 'Genome Quality Flags', 'Isolation Source',\n",
       "       'Isolation Comments', 'Collection Date', 'Collection Year', 'Season',\n",
       "       'Isolation Country', 'Geographic Group', 'Geographic Location',\n",
       "       'Other Environmental', 'Host Name', 'Host Common Name', 'Host Gender',\n",
       "       'Host Age', 'Host Health', 'Host Group', 'Lab Host', 'Passage',\n",
       "       'Other Clinical', 'Additional Metadata', 'Comments', 'Date Inserted',\n",
       "       'Date Modified'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Genome ID</th>\n",
       "      <th>Genome Name</th>\n",
       "      <th>Other Names</th>\n",
       "      <th>NCBI Taxon ID</th>\n",
       "      <th>Taxon Lineage IDs</th>\n",
       "      <th>Taxon Lineage Names</th>\n",
       "      <th>Superkingdom</th>\n",
       "      <th>Kingdom</th>\n",
       "      <th>Phylum</th>\n",
       "      <th>Class</th>\n",
       "      <th>...</th>\n",
       "      <th>Host Age</th>\n",
       "      <th>Host Health</th>\n",
       "      <th>Host Group</th>\n",
       "      <th>Lab Host</th>\n",
       "      <th>Passage</th>\n",
       "      <th>Other Clinical</th>\n",
       "      <th>Additional Metadata</th>\n",
       "      <th>Comments</th>\n",
       "      <th>Date Inserted</th>\n",
       "      <th>Date Modified</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>11053.10000</td>\n",
       "      <td>Dengue virus 1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11053</td>\n",
       "      <td>10239;2559587;2732396;2732406;2732462;2732545;...</td>\n",
       "      <td>Viruses;Riboviria;Orthornavirae;Kitrinoviricot...</td>\n",
       "      <td>Viruses</td>\n",
       "      <td>Orthornavirae</td>\n",
       "      <td>Kitrinoviricota</td>\n",
       "      <td>Flasuviricetes</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020-12-26T14:19:20.086Z</td>\n",
       "      <td>2020-12-26T14:19:20.086Z</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>11053.10000</td>\n",
       "      <td>Dengue virus 1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11053</td>\n",
       "      <td>10239;2559587;2732396;2732406;2732462;2732545;...</td>\n",
       "      <td>Viruses;Riboviria;Orthornavirae;Kitrinoviricot...</td>\n",
       "      <td>Viruses</td>\n",
       "      <td>Orthornavirae</td>\n",
       "      <td>Kitrinoviricota</td>\n",
       "      <td>Flasuviricetes</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020-12-26T19:23:35.931Z</td>\n",
       "      <td>2020-12-26T19:23:35.931Z</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>11053.10000</td>\n",
       "      <td>Dengue virus 1 Bali03</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11053</td>\n",
       "      <td>10239;2559587;2732396;2732406;2732462;2732545;...</td>\n",
       "      <td>Viruses;Riboviria;Orthornavirae;Kitrinoviricot...</td>\n",
       "      <td>Viruses</td>\n",
       "      <td>Orthornavirae</td>\n",
       "      <td>Kitrinoviricota</td>\n",
       "      <td>Flasuviricetes</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020-12-27T20:19:28.456Z</td>\n",
       "      <td>2020-12-27T20:19:28.456Z</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>11053.10000</td>\n",
       "      <td>Dengue virus 1 D1/Myanmar/1508aTw</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11053</td>\n",
       "      <td>10239;2559587;2732396;2732406;2732462;2732545;...</td>\n",
       "      <td>Viruses;Riboviria;Orthornavirae;Kitrinoviricot...</td>\n",
       "      <td>Viruses</td>\n",
       "      <td>Orthornavirae</td>\n",
       "      <td>Kitrinoviricota</td>\n",
       "      <td>Flasuviricetes</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2021-01-21T04:02:27.919Z</td>\n",
       "      <td>2021-01-21T04:02:27.919Z</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>11053.10001</td>\n",
       "      <td>Dengue virus 1 D1/Vietnam/1508aTw</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11053</td>\n",
       "      <td>10239;2559587;2732396;2732406;2732462;2732545;...</td>\n",
       "      <td>Viruses;Riboviria;Orthornavirae;Kitrinoviricot...</td>\n",
       "      <td>Viruses</td>\n",
       "      <td>Orthornavirae</td>\n",
       "      <td>Kitrinoviricota</td>\n",
       "      <td>Flasuviricetes</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2021-01-21T04:02:28.859Z</td>\n",
       "      <td>2021-01-21T04:02:28.859Z</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 87 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Genome ID                        Genome Name  Other Names  \\\n",
       "24  11053.10000                     Dengue virus 1          NaN   \n",
       "25  11053.10000                     Dengue virus 1          NaN   \n",
       "26  11053.10000              Dengue virus 1 Bali03          NaN   \n",
       "27  11053.10000  Dengue virus 1 D1/Myanmar/1508aTw          NaN   \n",
       "28  11053.10001  Dengue virus 1 D1/Vietnam/1508aTw          NaN   \n",
       "\n",
       "    NCBI Taxon ID                                  Taxon Lineage IDs  \\\n",
       "24          11053  10239;2559587;2732396;2732406;2732462;2732545;...   \n",
       "25          11053  10239;2559587;2732396;2732406;2732462;2732545;...   \n",
       "26          11053  10239;2559587;2732396;2732406;2732462;2732545;...   \n",
       "27          11053  10239;2559587;2732396;2732406;2732462;2732545;...   \n",
       "28          11053  10239;2559587;2732396;2732406;2732462;2732545;...   \n",
       "\n",
       "                                  Taxon Lineage Names Superkingdom  \\\n",
       "24  Viruses;Riboviria;Orthornavirae;Kitrinoviricot...      Viruses   \n",
       "25  Viruses;Riboviria;Orthornavirae;Kitrinoviricot...      Viruses   \n",
       "26  Viruses;Riboviria;Orthornavirae;Kitrinoviricot...      Viruses   \n",
       "27  Viruses;Riboviria;Orthornavirae;Kitrinoviricot...      Viruses   \n",
       "28  Viruses;Riboviria;Orthornavirae;Kitrinoviricot...      Viruses   \n",
       "\n",
       "          Kingdom           Phylum           Class  ... Host Age Host Health  \\\n",
       "24  Orthornavirae  Kitrinoviricota  Flasuviricetes  ...      NaN         NaN   \n",
       "25  Orthornavirae  Kitrinoviricota  Flasuviricetes  ...      NaN         NaN   \n",
       "26  Orthornavirae  Kitrinoviricota  Flasuviricetes  ...      NaN         NaN   \n",
       "27  Orthornavirae  Kitrinoviricota  Flasuviricetes  ...      NaN         NaN   \n",
       "28  Orthornavirae  Kitrinoviricota  Flasuviricetes  ...      NaN         NaN   \n",
       "\n",
       "   Host Group Lab Host Passage Other Clinical Additional Metadata  Comments  \\\n",
       "24      Human      NaN     NaN            NaN                 NaN       NaN   \n",
       "25      Human      NaN     NaN            NaN                 NaN       NaN   \n",
       "26        NaN      NaN     NaN            NaN                 NaN       NaN   \n",
       "27      Human      NaN     NaN            NaN                 NaN       NaN   \n",
       "28      Human      NaN     NaN            NaN                 NaN       NaN   \n",
       "\n",
       "               Date Inserted             Date Modified  \n",
       "24  2020-12-26T14:19:20.086Z  2020-12-26T14:19:20.086Z  \n",
       "25  2020-12-26T19:23:35.931Z  2020-12-26T19:23:35.931Z  \n",
       "26  2020-12-27T20:19:28.456Z  2020-12-27T20:19:28.456Z  \n",
       "27  2021-01-21T04:02:27.919Z  2021-01-21T04:02:27.919Z  \n",
       "28  2021-01-21T04:02:28.859Z  2021-01-21T04:02:28.859Z  \n",
       "\n",
       "[5 rows x 87 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# filtering for dengue virus seqs only by using the Species = Dengue virus\n",
    "\n",
    "data = data[data[\"Species\"] == \"Dengue virus\"]\n",
    "data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# deleting unnecesssary cols\n",
    "data = data.drop(columns = [\"Other Names\", \"NCBI Taxon ID\", \"Taxon Lineage IDs\", \"Taxon Lineage Names\", \n",
    "                        \"Superkingdom\", \"Kingdom\", \"Phylum\", \"Class\", \"Order\", \"Order\", \"Family\", \"Genus\",\n",
    "                        \"Genome Status\", \"Strain\", \"Serovar\", \"Biovar\", \"Pathovar\", \"MLST\", \"Segment\", \"Subtype\", \n",
    "                        \"H_type\", \"N_type\", \"H1 Clade Global\", \"H1 Clade US\", \"H5 Clade\", \"pH1N1-like\", \n",
    "                        \"Lineage\", \"Clade\", \"Subclade\", \"Other Typing\", \"Culture Collection\", \n",
    "                        \"Type Strain\", \"Reference\", \"Genome Quality\", \"Publication\", \"Authors\", \n",
    "                        \"BioProject Accession\", \"BioSample Accession\", \"Assembly Accession\", \"SRA Accession\", \n",
    "                        \"Sequencing Center\", \"Sequencing Status\", \"Sequencing Depth\", \"Assembly Method\",\n",
    "                        \"Chromosome\", \"Plasmids\", \"Contigs\", \"GC Content\", \"Contig L50\", \"Contig N50\", \"TRNA\",\n",
    "                        \"RRNA\", \"Mat Peptide\", \"CDS\", 'Coarse Consistency', \"Fine Consistency\",\n",
    "                        \"CheckM Contamination\", \"CheckM Completeness\", \"Genome Quality Flags\", \n",
    "                        \"Isolation Source\", \"Isolation Comments\", \"Collection Year\", \"Season\", \n",
    "                        \"Geographic Location\", \"Other Environmental\", \"Host Gender\", \"Host Age\", \n",
    "                        \"Host Health\", \"Host Group\", \"Lab Host\", \"Passage\", \"Other Clinical\", \n",
    "                        \"Additional Metadata\", \"Comments\", \"Date Inserted\", \"Date Modified\", \n",
    "                        \"Sequencing Platform\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Genome ID</th>\n",
       "      <th>Genome Name</th>\n",
       "      <th>Species</th>\n",
       "      <th>Completion Date</th>\n",
       "      <th>GenBank Accessions</th>\n",
       "      <th>Size</th>\n",
       "      <th>Collection Date</th>\n",
       "      <th>Isolation Country</th>\n",
       "      <th>Geographic Group</th>\n",
       "      <th>Host Name</th>\n",
       "      <th>Host Common Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>11053.10000</td>\n",
       "      <td>Dengue virus 1</td>\n",
       "      <td>Dengue virus</td>\n",
       "      <td>2020-11-10T00:00:00Z</td>\n",
       "      <td>MW208042</td>\n",
       "      <td>9592</td>\n",
       "      <td>2018-10-01</td>\n",
       "      <td>Brazil</td>\n",
       "      <td>South America</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>11053.10000</td>\n",
       "      <td>Dengue virus 1</td>\n",
       "      <td>Dengue virus</td>\n",
       "      <td>2014-10-30T00:00:00Z</td>\n",
       "      <td>KC692509</td>\n",
       "      <td>10735</td>\n",
       "      <td>2009-05-05</td>\n",
       "      <td>Argentina</td>\n",
       "      <td>South America</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>11053.10000</td>\n",
       "      <td>Dengue virus 1 Bali03</td>\n",
       "      <td>Dengue virus</td>\n",
       "      <td>2016-07-26T00:00:00Z</td>\n",
       "      <td>AY705999</td>\n",
       "      <td>1276</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Indonesia</td>\n",
       "      <td>Asia</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>11053.10000</td>\n",
       "      <td>Dengue virus 1 D1/Myanmar/1508aTw</td>\n",
       "      <td>Dengue virus</td>\n",
       "      <td>2018-08-17T00:00:00Z</td>\n",
       "      <td>MG894871</td>\n",
       "      <td>1485</td>\n",
       "      <td>2015-08-19</td>\n",
       "      <td>Taiwan</td>\n",
       "      <td>Asia</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>11053.10001</td>\n",
       "      <td>Dengue virus 1 D1/Vietnam/1508aTw</td>\n",
       "      <td>Dengue virus</td>\n",
       "      <td>2018-08-17T00:00:00Z</td>\n",
       "      <td>MG894872</td>\n",
       "      <td>1485</td>\n",
       "      <td>2015-08-08</td>\n",
       "      <td>Taiwan</td>\n",
       "      <td>Asia</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Human</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Genome ID                        Genome Name       Species  \\\n",
       "24  11053.10000                     Dengue virus 1  Dengue virus   \n",
       "25  11053.10000                     Dengue virus 1  Dengue virus   \n",
       "26  11053.10000              Dengue virus 1 Bali03  Dengue virus   \n",
       "27  11053.10000  Dengue virus 1 D1/Myanmar/1508aTw  Dengue virus   \n",
       "28  11053.10001  Dengue virus 1 D1/Vietnam/1508aTw  Dengue virus   \n",
       "\n",
       "         Completion Date GenBank Accessions   Size Collection Date  \\\n",
       "24  2020-11-10T00:00:00Z           MW208042   9592      2018-10-01   \n",
       "25  2014-10-30T00:00:00Z           KC692509  10735      2009-05-05   \n",
       "26  2016-07-26T00:00:00Z           AY705999   1276             NaN   \n",
       "27  2018-08-17T00:00:00Z           MG894871   1485      2015-08-19   \n",
       "28  2018-08-17T00:00:00Z           MG894872   1485      2015-08-08   \n",
       "\n",
       "   Isolation Country Geographic Group     Host Name Host Common Name  \n",
       "24            Brazil    South America  Homo sapiens            Human  \n",
       "25         Argentina    South America  Homo sapiens            Human  \n",
       "26         Indonesia             Asia           NaN              NaN  \n",
       "27            Taiwan             Asia  Homo sapiens            Human  \n",
       "28            Taiwan             Asia  Homo sapiens            Human  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Int64Index: 42790 entries, 24 to 43863\n",
      "Data columns (total 11 columns):\n",
      " #   Column              Non-Null Count  Dtype  \n",
      "---  ------              --------------  -----  \n",
      " 0   Genome ID           42790 non-null  float64\n",
      " 1   Genome Name         42790 non-null  object \n",
      " 2   Species             42790 non-null  object \n",
      " 3   Completion Date     42790 non-null  object \n",
      " 4   GenBank Accessions  42790 non-null  object \n",
      " 5   Size                42790 non-null  int64  \n",
      " 6   Collection Date     37199 non-null  object \n",
      " 7   Isolation Country   39758 non-null  object \n",
      " 8   Geographic Group    39753 non-null  object \n",
      " 9   Host Name           35585 non-null  object \n",
      " 10  Host Common Name    36351 non-null  object \n",
      "dtypes: float64(1), int64(1), object(9)\n",
      "memory usage: 3.9+ MB\n"
     ]
    }
   ],
   "source": [
    "data.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# optimizing memory\n",
    "data[\"Species\"] = data[\"Species\"].astype(\"category\")\n",
    "data[\"Geographic Group\"] = data[\"Geographic Group\"].astype(\"category\")\n",
    "data[\"Isolation Country\"] = data[\"Isolation Country\"].astype(\"category\")\n",
    "data[\"Host Name\"] = data[\"Host Name\"].astype(\"category\")\n",
    "data[\"Host Common Name\"] = data[\"Host Common Name\"].astype(\"category\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Int64Index: 42790 entries, 24 to 43863\n",
      "Data columns (total 11 columns):\n",
      " #   Column              Non-Null Count  Dtype   \n",
      "---  ------              --------------  -----   \n",
      " 0   Genome ID           42790 non-null  float64 \n",
      " 1   Genome Name         42790 non-null  object  \n",
      " 2   Species             42790 non-null  category\n",
      " 3   Completion Date     42790 non-null  object  \n",
      " 4   GenBank Accessions  42790 non-null  object  \n",
      " 5   Size                42790 non-null  int64   \n",
      " 6   Collection Date     37199 non-null  object  \n",
      " 7   Isolation Country   39758 non-null  category\n",
      " 8   Geographic Group    39753 non-null  category\n",
      " 9   Host Name           35585 non-null  category\n",
      " 10  Host Common Name    36351 non-null  category\n",
      "dtypes: category(5), float64(1), int64(1), object(4)\n",
      "memory usage: 2.6+ MB\n"
     ]
    }
   ],
   "source": [
    "data.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# renaming columns\n",
    "columns_map = {\"GenBank Accessions\": \"GenBank Accession\",\n",
    "              \"Isolation Country\": \"country\",\n",
    "              \"Collection Date\": \"collection_date\"}\n",
    "data.rename(columns = columns_map, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Genome ID</th>\n",
       "      <th>Genome Name</th>\n",
       "      <th>Species</th>\n",
       "      <th>Completion Date</th>\n",
       "      <th>GenBank Accession</th>\n",
       "      <th>Size</th>\n",
       "      <th>collection_date</th>\n",
       "      <th>country</th>\n",
       "      <th>Geographic Group</th>\n",
       "      <th>Host Name</th>\n",
       "      <th>Host Common Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>11053.10000</td>\n",
       "      <td>Dengue virus 1</td>\n",
       "      <td>Dengue virus</td>\n",
       "      <td>2020-11-10T00:00:00Z</td>\n",
       "      <td>MW208042</td>\n",
       "      <td>9592</td>\n",
       "      <td>2018-10-01</td>\n",
       "      <td>Brazil</td>\n",
       "      <td>South America</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>11053.10000</td>\n",
       "      <td>Dengue virus 1</td>\n",
       "      <td>Dengue virus</td>\n",
       "      <td>2014-10-30T00:00:00Z</td>\n",
       "      <td>KC692509</td>\n",
       "      <td>10735</td>\n",
       "      <td>2009-05-05</td>\n",
       "      <td>Argentina</td>\n",
       "      <td>South America</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>11053.10000</td>\n",
       "      <td>Dengue virus 1 Bali03</td>\n",
       "      <td>Dengue virus</td>\n",
       "      <td>2016-07-26T00:00:00Z</td>\n",
       "      <td>AY705999</td>\n",
       "      <td>1276</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Indonesia</td>\n",
       "      <td>Asia</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>11053.10000</td>\n",
       "      <td>Dengue virus 1 D1/Myanmar/1508aTw</td>\n",
       "      <td>Dengue virus</td>\n",
       "      <td>2018-08-17T00:00:00Z</td>\n",
       "      <td>MG894871</td>\n",
       "      <td>1485</td>\n",
       "      <td>2015-08-19</td>\n",
       "      <td>Taiwan</td>\n",
       "      <td>Asia</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>11053.10001</td>\n",
       "      <td>Dengue virus 1 D1/Vietnam/1508aTw</td>\n",
       "      <td>Dengue virus</td>\n",
       "      <td>2018-08-17T00:00:00Z</td>\n",
       "      <td>MG894872</td>\n",
       "      <td>1485</td>\n",
       "      <td>2015-08-08</td>\n",
       "      <td>Taiwan</td>\n",
       "      <td>Asia</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Human</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Genome ID                        Genome Name       Species  \\\n",
       "24  11053.10000                     Dengue virus 1  Dengue virus   \n",
       "25  11053.10000                     Dengue virus 1  Dengue virus   \n",
       "26  11053.10000              Dengue virus 1 Bali03  Dengue virus   \n",
       "27  11053.10000  Dengue virus 1 D1/Myanmar/1508aTw  Dengue virus   \n",
       "28  11053.10001  Dengue virus 1 D1/Vietnam/1508aTw  Dengue virus   \n",
       "\n",
       "         Completion Date GenBank Accession   Size collection_date    country  \\\n",
       "24  2020-11-10T00:00:00Z          MW208042   9592      2018-10-01     Brazil   \n",
       "25  2014-10-30T00:00:00Z          KC692509  10735      2009-05-05  Argentina   \n",
       "26  2016-07-26T00:00:00Z          AY705999   1276             NaN  Indonesia   \n",
       "27  2018-08-17T00:00:00Z          MG894871   1485      2015-08-19     Taiwan   \n",
       "28  2018-08-17T00:00:00Z          MG894872   1485      2015-08-08     Taiwan   \n",
       "\n",
       "   Geographic Group     Host Name Host Common Name  \n",
       "24    South America  Homo sapiens            Human  \n",
       "25    South America  Homo sapiens            Human  \n",
       "26             Asia           NaN              NaN  \n",
       "27             Asia  Homo sapiens            Human  \n",
       "28             Asia  Homo sapiens            Human  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Standardizing country names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# dictionary to standardize African country names\n",
    "\n",
    "standardize_africa_countries = {\"Peopleâ€™s Democratic Republic of Algeria\": \"Algeria\", \n",
    "                                \"Republic of Angola\": \"Angola\", \"Republic of Benin\": \"Benin\", \n",
    "                                \"Republic of Botswana\": \"Botswana\", \"Republic of Burundi\": \"Burundi\", \n",
    "                                \"Republic of Cameroon\": \"Cameroon\", \"Republic of Cabo Verde\": \"Cabo Verde\",\n",
    "                                \"Cape Verde\": \"Cabo Verde\",\"Republic of Chad\": \"Chad\", \n",
    "                                \"Union of the Comoros\": \"Comoros\", \"DR Congo\": \"Democratic Republic of the Congo\",\n",
    "                                \"Republic of Djibouti\": \"Djibouti\", \"Arab Republic of Egypt\": \"Egypt\",\n",
    "                                \"Republic of Equatorial Guinea\": \"Equatorial Guinea\", \n",
    "                                \"State of Eritrea\": \"Eritrea\", \"Kingdom of eSwatini\": \"eSwatini\", \n",
    "                                \"Swaziland\": \"eSwatini\", \"Federal Democratic Republic of Ethiopia\": \"Ethiopia\", \n",
    "                                \"Gabonese Republic\": \"Gabon\", \"Republic of Gambia\": \"Gambia\", \n",
    "                                \"Republic of Ghana\": \"Ghana\", \"Republic of Guinea\": \"Guinea\", \n",
    "                                \"Republic of Guinea-Bissau\": \"Guinea-Bissau\",\n",
    "                                \"Republic of CÃ´te d\\\"Ivoire\": \"Cote d\\\"Ivoire\", \"Ivory Coast\": \"Cote d\\\"Ivoire\",\n",
    "                                \"Republic of Kenya\": \"Kenya\", \"Kingdom of Lesotho\": \"Lesotho\",\n",
    "                                \"Republic of Liberia\": \"Liberia\", \"State of Libya\": \"Libya\",\n",
    "                                \"Republic of Madagascar\": \"Madagascar\", \"Republic of Malawi\": \"Malawi\",\n",
    "                                \"Republic of Mali\": \"Mali\", \"Islamic Republic of Mauritania\": \"Mauritania\",\n",
    "                                \"Republic of Mauritius\": \"Mauritius\", \"Kingdom of Morocco\": \"Morocco\",\n",
    "                                \"Republic of Mozambique\": \"Mozambique\", \"Republic of Namibia\": \"Namibia\",\n",
    "                                \"Republic of Niger\": \"Niger\", \"Federal Republic of Nigeria\": \"Nigeria\",\n",
    "                                \"Congo\":\"Republic of the Congo\", \"La Reunion\": \"Reunion\",\n",
    "                                \"Republic of Rwanda\": \"Rwanda\",\n",
    "                                \"Democratic Republic of SÃ£o TomÃ© and Principe\": \"SÃ£o TomÃ© and Principe\",\n",
    "                                \"Republic of Senegal\": \"Senegal\", \"Republic of Seychelles\": \"Seychelles\",\n",
    "                                \"Republic of Sierra Leone\": \"Sierra Leone\",\n",
    "                                \"Federal Republic of Somalia\": \"Somalia\", \n",
    "                                \"Republic of South Africa\": \"South Africa\",\n",
    "                                \"Republic of South Sudan\": \"South Sudan\", \"Republic of the Sudan\": \"Sudan\",\n",
    "                                \"United Republic of Tanzania\": \"Tanzania\", \"Togolese\": \"Togo\",\n",
    "                                \"Togolese Republic\": \"Togo\", \"Republic of Tunisia\": \"Tunisia\",\n",
    "                                \"Republic of Uganda\": \"Uganda\", \"Republic of Zambia\": \"Zambia\",\n",
    "                                \"Republic of Zimbabwe\": \"Zimbabwe\"}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# standardizing country names\n",
    "\n",
    "data.replace({\"country\": standardize_africa_countries}, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Genome ID</th>\n",
       "      <th>Genome Name</th>\n",
       "      <th>Species</th>\n",
       "      <th>Completion Date</th>\n",
       "      <th>GenBank Accession</th>\n",
       "      <th>Size</th>\n",
       "      <th>collection_date</th>\n",
       "      <th>country</th>\n",
       "      <th>Geographic Group</th>\n",
       "      <th>Host Name</th>\n",
       "      <th>Host Common Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Genome ID, Genome Name, Species, Completion Date, GenBank Accession, Size, collection_date, country, Geographic Group, Host Name, Host Common Name]\n",
       "Index: []"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# checking if African countries are in data set\n",
    "\n",
    "data[data[\"country\"].isin([\"Central African Republic\"])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "african_countries = [\"Algeria\", \"Angola\", \"Benin\", \"Botswana\", \"Burkina Faso\", \"Burundi\", \"Cameroon\", \n",
    "                     \"Cabo Verde\", \"Central African Republic\", \"Chad\", \"Comoros\", \n",
    "                     \"Democratic Republic of the Congo\", \"Djibouti\", \"Egypt\", \"Equatorial Guinea\", \"Eritrea\", \n",
    "                     \"eSwatini\", \"Ethiopia\", \"Gabon\", \"Gambia\", \"Ghana\", \"Guinea\", \"Guinea-Bissau\", \n",
    "                     \"Cote d\\\"Ivoire\", \"Kenya\", \"Lesotho\", \"Liberia\", \"Libya\", \"Madagascar\", \"Malawi\", \n",
    "                     \"Mali\", \"Mauritania\", \"Mauritius\", \"Morocco\", \"Mozambique\", \"Namibia\",\"Niger\", \"Nigeria\", \n",
    "                     \"Republic of the Congo\", \"Reunion\", \"Rwanda\", \"SÃ£o TomÃ© and Principe\", \"Senegal\", \n",
    "                     \"Seychelles\", \"Sierra Leone\", \"Somalia\", \"South Africa\", \"South Sudan\", \"Sudan\", \n",
    "                     \"Tanzania\", \"Togo\", \"Tunisia\", \"Uganda\", \"Zambia\", \"Zimbabwe\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "55"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(african_countries)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# filtering for African countries only\n",
    "\n",
    "df_africa = data[data[\"country\"].isin(african_countries)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Genome ID</th>\n",
       "      <th>Genome Name</th>\n",
       "      <th>Species</th>\n",
       "      <th>Completion Date</th>\n",
       "      <th>GenBank Accession</th>\n",
       "      <th>Size</th>\n",
       "      <th>collection_date</th>\n",
       "      <th>country</th>\n",
       "      <th>Geographic Group</th>\n",
       "      <th>Host Name</th>\n",
       "      <th>Host Common Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1096</th>\n",
       "      <td>11053.10967</td>\n",
       "      <td>Dengue virus 1 SF729VI06</td>\n",
       "      <td>Dengue virus</td>\n",
       "      <td>2016-07-24T00:00:00Z</td>\n",
       "      <td>FJ502874</td>\n",
       "      <td>216</td>\n",
       "      <td>2005</td>\n",
       "      <td>Madagascar</td>\n",
       "      <td>Africa</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1112</th>\n",
       "      <td>11053.10981</td>\n",
       "      <td>Dengue virus 1 CNM664VI06</td>\n",
       "      <td>Dengue virus</td>\n",
       "      <td>2016-07-24T00:00:00Z</td>\n",
       "      <td>FJ502864</td>\n",
       "      <td>216</td>\n",
       "      <td>2006</td>\n",
       "      <td>Cameroon</td>\n",
       "      <td>Africa</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1113</th>\n",
       "      <td>11053.10982</td>\n",
       "      <td>Dengue virus 1 CNM399VI07</td>\n",
       "      <td>Dengue virus</td>\n",
       "      <td>2016-07-24T00:00:00Z</td>\n",
       "      <td>FJ502863</td>\n",
       "      <td>216</td>\n",
       "      <td>2007</td>\n",
       "      <td>Nigeria</td>\n",
       "      <td>Africa</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1157</th>\n",
       "      <td>11053.11020</td>\n",
       "      <td>Dengue virus 1 AO/DB136/2013</td>\n",
       "      <td>Dengue virus</td>\n",
       "      <td>2014-11-23T00:00:00Z</td>\n",
       "      <td>KM277614</td>\n",
       "      <td>1485</td>\n",
       "      <td>2013</td>\n",
       "      <td>Angola</td>\n",
       "      <td>Africa</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1160</th>\n",
       "      <td>11053.11023</td>\n",
       "      <td>Dengue virus 1 AO/DB135/2013</td>\n",
       "      <td>Dengue virus</td>\n",
       "      <td>2014-11-23T00:00:00Z</td>\n",
       "      <td>KM277613</td>\n",
       "      <td>1485</td>\n",
       "      <td>2013</td>\n",
       "      <td>Angola</td>\n",
       "      <td>Africa</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Human</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Genome ID                   Genome Name       Species  \\\n",
       "1096  11053.10967      Dengue virus 1 SF729VI06  Dengue virus   \n",
       "1112  11053.10981     Dengue virus 1 CNM664VI06  Dengue virus   \n",
       "1113  11053.10982     Dengue virus 1 CNM399VI07  Dengue virus   \n",
       "1157  11053.11020  Dengue virus 1 AO/DB136/2013  Dengue virus   \n",
       "1160  11053.11023  Dengue virus 1 AO/DB135/2013  Dengue virus   \n",
       "\n",
       "           Completion Date GenBank Accession  Size collection_date  \\\n",
       "1096  2016-07-24T00:00:00Z          FJ502874   216            2005   \n",
       "1112  2016-07-24T00:00:00Z          FJ502864   216            2006   \n",
       "1113  2016-07-24T00:00:00Z          FJ502863   216            2007   \n",
       "1157  2014-11-23T00:00:00Z          KM277614  1485            2013   \n",
       "1160  2014-11-23T00:00:00Z          KM277613  1485            2013   \n",
       "\n",
       "         country Geographic Group     Host Name Host Common Name  \n",
       "1096  Madagascar           Africa  Homo sapiens            Human  \n",
       "1112    Cameroon           Africa  Homo sapiens            Human  \n",
       "1113     Nigeria           Africa  Homo sapiens            Human  \n",
       "1157      Angola           Africa  Homo sapiens            Human  \n",
       "1160      Angola           Africa  Homo sapiens            Human  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_africa.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "877"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df_africa)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_133/2721846101.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_africa.drop_duplicates(subset = [\"GenBank Accession\"], keep = \"first\", inplace=True)\n"
     ]
    }
   ],
   "source": [
    "# dropping duplicates from the GenBank Accessions col\n",
    "\n",
    "df_africa.drop_duplicates(subset = [\"GenBank Accession\"], keep = \"first\", inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "873"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df_africa)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "# exporting africa dataset\n",
    "\n",
    "df_africa.to_csv(\"Africa_dengue.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "# grabbing all the GenBank accession numbers \n",
    "\n",
    "genbank = df_africa[\"GenBank Accession\"].to_list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "# writing to .txt file\n",
    "\n",
    "with open(\"genbank_accession.txt\", \"w\") as f:\n",
    "    f.write(\"\\n\".join(genbank))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<a href='Africa_dengue.csv' target='_blank'>Africa_dengue.csv</a><br>"
      ],
      "text/plain": [
       "/home/jupyter/VEME 2023 Pathogen Dashboards/edit/Africa_dengue.csv"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<a href='genbank_accession.txt' target='_blank'>genbank_accession.txt</a><br>"
      ],
      "text/plain": [
       "/home/jupyter/VEME 2023 Pathogen Dashboards/edit/genbank_accession.txt"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Downloading files\n",
    "dataset_file = \"Africa_dengue.csv\"\n",
    "genbank_file = \"genbank_accession.txt\"\n",
    "\n",
    "display(FileLink(dataset_file))\n",
    "display(FileLink(genbank_file))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Running Genome Detective\n",
    "- Downloading all the fasta files for Africa from Genbank\n",
    "- Running through the dengue typing tool on GD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sequences downloaded and saved to dengue.fasta\n"
     ]
    }
   ],
   "source": [
    "# downloading fasta files from NCBI \n",
    "\n",
    "from Bio import Entrez\n",
    "\n",
    "Entrez.email = \"your_email\"\n",
    "\n",
    "genbank_ids = genbank  \n",
    "\n",
    "db = \"nucleotide\"  \n",
    "rettype = \"fasta\"  \n",
    "retmode = \"text\"  \n",
    "\n",
    "sequences = []  \n",
    "\n",
    "for genbank_id in genbank_ids:\n",
    "    handle = Entrez.efetch(db=db, id=genbank_id, rettype=rettype, retmode=retmode)\n",
    "    sequence = handle.read()  \n",
    "    handle.close()\n",
    "    sequences.append(sequence)\n",
    "    \n",
    "output_file = \"dengue.fasta\"  \n",
    "\n",
    "with open(output_file, \"w\") as file:\n",
    "    file.write(\"\".join(sequences))\n",
    "\n",
    "print(f\"Sequences downloaded and saved to {output_file}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<a href='dengue.fasta' target='_blank'>dengue.fasta</a><br>"
      ],
      "text/plain": [
       "/home/jupyter/VEME 2023 Pathogen Dashboards/edit/dengue.fasta"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Downloading fasta file\n",
    "FileLink(\"dengue.fasta\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### RUN ON GENOME DETECTIVE\n",
    "\n",
    "- Drop the text file with all the fasta sequences into the Genome Detective Dengue Virus Typing Tool\n",
    "- https://www.genomedetective.com/app/typingtool/dengue/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data process for data from Genome Detective (GD)\n",
    "- Reading in and editing the GD data\n",
    "- Merging cleaned and subsetted VIPR data with newly run GD data\n",
    "- Changing the serotypes and genotypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "GD_data = pd.read_csv(\"gs://fc-c4948adf-e9ec-4849-86cd-eee5235c0bae/Backup Files/results.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>assignment</th>\n",
       "      <th>length</th>\n",
       "      <th>begin</th>\n",
       "      <th>end</th>\n",
       "      <th>species</th>\n",
       "      <th>species-score</th>\n",
       "      <th>type</th>\n",
       "      <th>type support</th>\n",
       "      <th>subtype</th>\n",
       "      <th>subtype support</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>FJ502874.1</td>\n",
       "      <td>1IV</td>\n",
       "      <td>216</td>\n",
       "      <td>2196.0</td>\n",
       "      <td>2411.0</td>\n",
       "      <td>dengue virus type 1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DENV-1 Genotype IV</td>\n",
       "      <td>86.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FJ502864.1</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>216</td>\n",
       "      <td>2196.0</td>\n",
       "      <td>2411.0</td>\n",
       "      <td>dengue virus type 1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Could not assign</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FJ502863.1</td>\n",
       "      <td>1V</td>\n",
       "      <td>216</td>\n",
       "      <td>2196.0</td>\n",
       "      <td>2411.0</td>\n",
       "      <td>dengue virus type 1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DENV-1 Genotype V</td>\n",
       "      <td>80.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>KM277614.1</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>1485</td>\n",
       "      <td>935.0</td>\n",
       "      <td>2419.0</td>\n",
       "      <td>dengue virus type 1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Could not assign</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>KM277613.1</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>1485</td>\n",
       "      <td>935.0</td>\n",
       "      <td>2419.0</td>\n",
       "      <td>dengue virus type 1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Could not assign</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         name  assignment  length   begin     end              species  \\\n",
       "0  FJ502874.1         1IV     216  2196.0  2411.0  dengue virus type 1   \n",
       "1  FJ502864.1  Unassigned     216  2196.0  2411.0  dengue virus type 1   \n",
       "2  FJ502863.1          1V     216  2196.0  2411.0  dengue virus type 1   \n",
       "3  KM277614.1  Unassigned    1485   935.0  2419.0  dengue virus type 1   \n",
       "4  KM277613.1  Unassigned    1485   935.0  2419.0  dengue virus type 1   \n",
       "\n",
       "   species-score                type  type support  subtype  subtype support  \n",
       "0            NaN  DENV-1 Genotype IV          86.0      NaN              NaN  \n",
       "1            NaN    Could not assign           NaN      NaN              NaN  \n",
       "2            NaN   DENV-1 Genotype V          80.0      NaN              NaN  \n",
       "3            NaN    Could not assign           NaN      NaN              NaN  \n",
       "4            NaN    Could not assign           NaN      NaN              NaN  "
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GD_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "# changing the names back to the original GenBank accession IDs\n",
    "# defining function to change this\n",
    "\n",
    "def fix_name(row):\n",
    "    name = row[0]\n",
    "    \n",
    "    if \".\" in name:\n",
    "        return name.split(\".\")[0]\n",
    "    \n",
    "    else:\n",
    "        return name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>assignment</th>\n",
       "      <th>length</th>\n",
       "      <th>begin</th>\n",
       "      <th>end</th>\n",
       "      <th>species</th>\n",
       "      <th>species-score</th>\n",
       "      <th>type</th>\n",
       "      <th>type support</th>\n",
       "      <th>subtype</th>\n",
       "      <th>subtype support</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>FJ502874</td>\n",
       "      <td>1IV</td>\n",
       "      <td>216</td>\n",
       "      <td>2196.0</td>\n",
       "      <td>2411.0</td>\n",
       "      <td>dengue virus type 1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DENV-1 Genotype IV</td>\n",
       "      <td>86.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FJ502864</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>216</td>\n",
       "      <td>2196.0</td>\n",
       "      <td>2411.0</td>\n",
       "      <td>dengue virus type 1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Could not assign</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FJ502863</td>\n",
       "      <td>1V</td>\n",
       "      <td>216</td>\n",
       "      <td>2196.0</td>\n",
       "      <td>2411.0</td>\n",
       "      <td>dengue virus type 1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DENV-1 Genotype V</td>\n",
       "      <td>80.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>KM277614</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>1485</td>\n",
       "      <td>935.0</td>\n",
       "      <td>2419.0</td>\n",
       "      <td>dengue virus type 1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Could not assign</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>KM277613</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>1485</td>\n",
       "      <td>935.0</td>\n",
       "      <td>2419.0</td>\n",
       "      <td>dengue virus type 1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Could not assign</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       name  assignment  length   begin     end              species  \\\n",
       "0  FJ502874         1IV     216  2196.0  2411.0  dengue virus type 1   \n",
       "1  FJ502864  Unassigned     216  2196.0  2411.0  dengue virus type 1   \n",
       "2  FJ502863          1V     216  2196.0  2411.0  dengue virus type 1   \n",
       "3  KM277614  Unassigned    1485   935.0  2419.0  dengue virus type 1   \n",
       "4  KM277613  Unassigned    1485   935.0  2419.0  dengue virus type 1   \n",
       "\n",
       "   species-score                type  type support  subtype  subtype support  \n",
       "0            NaN  DENV-1 Genotype IV          86.0      NaN              NaN  \n",
       "1            NaN    Could not assign           NaN      NaN              NaN  \n",
       "2            NaN   DENV-1 Genotype V          80.0      NaN              NaN  \n",
       "3            NaN    Could not assign           NaN      NaN              NaN  \n",
       "4            NaN    Could not assign           NaN      NaN              NaN  "
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# executing function\n",
    "\n",
    "GD_data[\"name\"] = GD_data.apply(fix_name, axis = \"columns\")\n",
    "GD_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>assignment</th>\n",
       "      <th>length</th>\n",
       "      <th>begin</th>\n",
       "      <th>end</th>\n",
       "      <th>species</th>\n",
       "      <th>species-score</th>\n",
       "      <th>type</th>\n",
       "      <th>type support</th>\n",
       "      <th>subtype</th>\n",
       "      <th>subtype support</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>FJ502874</td>\n",
       "      <td>1IV</td>\n",
       "      <td>216</td>\n",
       "      <td>2196.0</td>\n",
       "      <td>2411.0</td>\n",
       "      <td>dengue virus type 1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DENV-1 Genotype IV</td>\n",
       "      <td>86.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FJ502864</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>216</td>\n",
       "      <td>2196.0</td>\n",
       "      <td>2411.0</td>\n",
       "      <td>dengue virus type 1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Could not assign</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FJ502863</td>\n",
       "      <td>1V</td>\n",
       "      <td>216</td>\n",
       "      <td>2196.0</td>\n",
       "      <td>2411.0</td>\n",
       "      <td>dengue virus type 1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DENV-1 Genotype V</td>\n",
       "      <td>80.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>KM277614</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>1485</td>\n",
       "      <td>935.0</td>\n",
       "      <td>2419.0</td>\n",
       "      <td>dengue virus type 1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Could not assign</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>KM277613</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>1485</td>\n",
       "      <td>935.0</td>\n",
       "      <td>2419.0</td>\n",
       "      <td>dengue virus type 1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Could not assign</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>868</th>\n",
       "      <td>KM892495</td>\n",
       "      <td>2II</td>\n",
       "      <td>500</td>\n",
       "      <td>145.0</td>\n",
       "      <td>644.0</td>\n",
       "      <td>dengue virus type 2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DENV-2 Genotype II - Cosmopolitan</td>\n",
       "      <td>100.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>869</th>\n",
       "      <td>KU886233</td>\n",
       "      <td>2II</td>\n",
       "      <td>500</td>\n",
       "      <td>145.0</td>\n",
       "      <td>644.0</td>\n",
       "      <td>dengue virus type 2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DENV-2 Genotype II - Cosmopolitan</td>\n",
       "      <td>100.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>870</th>\n",
       "      <td>DQ016638</td>\n",
       "      <td>1I</td>\n",
       "      <td>306</td>\n",
       "      <td>2196.0</td>\n",
       "      <td>2501.0</td>\n",
       "      <td>dengue virus type 1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DENV-1 Genotype I</td>\n",
       "      <td>100.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>871</th>\n",
       "      <td>DQ016641</td>\n",
       "      <td>1I</td>\n",
       "      <td>306</td>\n",
       "      <td>2196.0</td>\n",
       "      <td>2501.0</td>\n",
       "      <td>dengue virus type 1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DENV-1 Genotype I</td>\n",
       "      <td>100.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>872</th>\n",
       "      <td>DQ016643</td>\n",
       "      <td>2V</td>\n",
       "      <td>306</td>\n",
       "      <td>2198.0</td>\n",
       "      <td>2503.0</td>\n",
       "      <td>dengue virus type 2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DENV-2 Genotype V - Asian I</td>\n",
       "      <td>98.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>873 rows Ã— 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         name  assignment  length   begin     end              species  \\\n",
       "0    FJ502874         1IV     216  2196.0  2411.0  dengue virus type 1   \n",
       "1    FJ502864  Unassigned     216  2196.0  2411.0  dengue virus type 1   \n",
       "2    FJ502863          1V     216  2196.0  2411.0  dengue virus type 1   \n",
       "3    KM277614  Unassigned    1485   935.0  2419.0  dengue virus type 1   \n",
       "4    KM277613  Unassigned    1485   935.0  2419.0  dengue virus type 1   \n",
       "..        ...         ...     ...     ...     ...                  ...   \n",
       "868  KM892495         2II     500   145.0   644.0  dengue virus type 2   \n",
       "869  KU886233         2II     500   145.0   644.0  dengue virus type 2   \n",
       "870  DQ016638          1I     306  2196.0  2501.0  dengue virus type 1   \n",
       "871  DQ016641          1I     306  2196.0  2501.0  dengue virus type 1   \n",
       "872  DQ016643          2V     306  2198.0  2503.0  dengue virus type 2   \n",
       "\n",
       "     species-score                               type  type support  subtype  \\\n",
       "0              NaN                 DENV-1 Genotype IV          86.0      NaN   \n",
       "1              NaN                   Could not assign           NaN      NaN   \n",
       "2              NaN                  DENV-1 Genotype V          80.0      NaN   \n",
       "3              NaN                   Could not assign           NaN      NaN   \n",
       "4              NaN                   Could not assign           NaN      NaN   \n",
       "..             ...                                ...           ...      ...   \n",
       "868            NaN  DENV-2 Genotype II - Cosmopolitan         100.0      NaN   \n",
       "869            NaN  DENV-2 Genotype II - Cosmopolitan         100.0      NaN   \n",
       "870            NaN                  DENV-1 Genotype I         100.0      NaN   \n",
       "871            NaN                  DENV-1 Genotype I         100.0      NaN   \n",
       "872            NaN        DENV-2 Genotype V - Asian I          98.0      NaN   \n",
       "\n",
       "     subtype support  \n",
       "0                NaN  \n",
       "1                NaN  \n",
       "2                NaN  \n",
       "3                NaN  \n",
       "4                NaN  \n",
       "..               ...  \n",
       "868              NaN  \n",
       "869              NaN  \n",
       "870              NaN  \n",
       "871              NaN  \n",
       "872              NaN  \n",
       "\n",
       "[873 rows x 11 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# dropping duplicates from the GD results using GenBank Accessions col\n",
    "\n",
    "GD_data.drop_duplicates(subset = [\"name\"], keep = \"first\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>length</th>\n",
       "      <th>species</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>FJ502874</td>\n",
       "      <td>216</td>\n",
       "      <td>dengue virus type 1</td>\n",
       "      <td>DENV-1 Genotype IV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FJ502864</td>\n",
       "      <td>216</td>\n",
       "      <td>dengue virus type 1</td>\n",
       "      <td>Could not assign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FJ502863</td>\n",
       "      <td>216</td>\n",
       "      <td>dengue virus type 1</td>\n",
       "      <td>DENV-1 Genotype V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>KM277614</td>\n",
       "      <td>1485</td>\n",
       "      <td>dengue virus type 1</td>\n",
       "      <td>Could not assign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>KM277613</td>\n",
       "      <td>1485</td>\n",
       "      <td>dengue virus type 1</td>\n",
       "      <td>Could not assign</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       name  length              species                type\n",
       "0  FJ502874     216  dengue virus type 1  DENV-1 Genotype IV\n",
       "1  FJ502864     216  dengue virus type 1    Could not assign\n",
       "2  FJ502863     216  dengue virus type 1   DENV-1 Genotype V\n",
       "3  KM277614    1485  dengue virus type 1    Could not assign\n",
       "4  KM277613    1485  dengue virus type 1    Could not assign"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# deleting unnecesssary cols\n",
    "\n",
    "GD_data = GD_data.drop(columns = [\"assignment\", \"begin\", \"end\", \"species-score\", \"type support\", \n",
    "                                \"subtype\", \"subtype support\"])\n",
    "GD_data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Genomic Dash Framework metadata template\n",
    "\n",
    "To fit on Genomic Dash Framework your csv file should have at least the following columns: \n",
    "- category_1: This should be the pathogen of your dashboard. For this tutorial we are going to use Dengue 1 (DENV-1)\n",
    "- category_2: The sub-categories you want to show, named as will be shown in the dashboard\n",
    "- date: date formatted\n",
    "- region: Continent of the data\n",
    "- country\n",
    "- province: optional"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "# renaming cols\n",
    "\n",
    "GD_data.rename(columns = {\"name\": \"GenBank Accession\"}, inplace = True)\n",
    "GD_data.rename(columns = {\"species\": \"category_1\"}, inplace = True)\n",
    "GD_data.rename(columns = {\"type\": \"category_2\"}, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GenBank Accession</th>\n",
       "      <th>length</th>\n",
       "      <th>category_1</th>\n",
       "      <th>category_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>FJ502874</td>\n",
       "      <td>216</td>\n",
       "      <td>dengue virus type 1</td>\n",
       "      <td>DENV-1 Genotype IV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FJ502864</td>\n",
       "      <td>216</td>\n",
       "      <td>dengue virus type 1</td>\n",
       "      <td>Could not assign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FJ502863</td>\n",
       "      <td>216</td>\n",
       "      <td>dengue virus type 1</td>\n",
       "      <td>DENV-1 Genotype V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>KM277614</td>\n",
       "      <td>1485</td>\n",
       "      <td>dengue virus type 1</td>\n",
       "      <td>Could not assign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>KM277613</td>\n",
       "      <td>1485</td>\n",
       "      <td>dengue virus type 1</td>\n",
       "      <td>Could not assign</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  GenBank Accession  length           category_1          category_2\n",
       "0          FJ502874     216  dengue virus type 1  DENV-1 Genotype IV\n",
       "1          FJ502864     216  dengue virus type 1    Could not assign\n",
       "2          FJ502863     216  dengue virus type 1   DENV-1 Genotype V\n",
       "3          KM277614    1485  dengue virus type 1    Could not assign\n",
       "4          KM277613    1485  dengue virus type 1    Could not assign"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GD_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# checking if the GD report has missing values in the assignment of the serotype (category_1)\n",
    "# need to fill them with unassigned\n",
    "\n",
    "GD_data[\"category_1\"].hasnans"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "# if true run...\n",
    "\n",
    "GD_data[\"category_1\"] = GD_data[\"category_1\"].fillna(\"Could not assign\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# checking if the GD report has missing values in the assignment of the genotype (category_2)\n",
    "# need to fill them with unassigned\n",
    "\n",
    "GD_data[\"category_2\"].hasnans"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "# if true run...\n",
    "\n",
    "GD_data[\"category_2\"] = GD_data[\"category_2\"].fillna(\"Could not assign\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Merging the VIPR and GD data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Genome ID</th>\n",
       "      <th>Genome Name</th>\n",
       "      <th>Species</th>\n",
       "      <th>Completion Date</th>\n",
       "      <th>GenBank Accession</th>\n",
       "      <th>Size</th>\n",
       "      <th>collection_date</th>\n",
       "      <th>country</th>\n",
       "      <th>Geographic Group</th>\n",
       "      <th>Host Name</th>\n",
       "      <th>Host Common Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1096</th>\n",
       "      <td>11053.10967</td>\n",
       "      <td>Dengue virus 1 SF729VI06</td>\n",
       "      <td>Dengue virus</td>\n",
       "      <td>2016-07-24T00:00:00Z</td>\n",
       "      <td>FJ502874</td>\n",
       "      <td>216</td>\n",
       "      <td>2005</td>\n",
       "      <td>Madagascar</td>\n",
       "      <td>Africa</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1112</th>\n",
       "      <td>11053.10981</td>\n",
       "      <td>Dengue virus 1 CNM664VI06</td>\n",
       "      <td>Dengue virus</td>\n",
       "      <td>2016-07-24T00:00:00Z</td>\n",
       "      <td>FJ502864</td>\n",
       "      <td>216</td>\n",
       "      <td>2006</td>\n",
       "      <td>Cameroon</td>\n",
       "      <td>Africa</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1113</th>\n",
       "      <td>11053.10982</td>\n",
       "      <td>Dengue virus 1 CNM399VI07</td>\n",
       "      <td>Dengue virus</td>\n",
       "      <td>2016-07-24T00:00:00Z</td>\n",
       "      <td>FJ502863</td>\n",
       "      <td>216</td>\n",
       "      <td>2007</td>\n",
       "      <td>Nigeria</td>\n",
       "      <td>Africa</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1157</th>\n",
       "      <td>11053.11020</td>\n",
       "      <td>Dengue virus 1 AO/DB136/2013</td>\n",
       "      <td>Dengue virus</td>\n",
       "      <td>2014-11-23T00:00:00Z</td>\n",
       "      <td>KM277614</td>\n",
       "      <td>1485</td>\n",
       "      <td>2013</td>\n",
       "      <td>Angola</td>\n",
       "      <td>Africa</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1160</th>\n",
       "      <td>11053.11023</td>\n",
       "      <td>Dengue virus 1 AO/DB135/2013</td>\n",
       "      <td>Dengue virus</td>\n",
       "      <td>2014-11-23T00:00:00Z</td>\n",
       "      <td>KM277613</td>\n",
       "      <td>1485</td>\n",
       "      <td>2013</td>\n",
       "      <td>Angola</td>\n",
       "      <td>Africa</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Human</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Genome ID                   Genome Name       Species  \\\n",
       "1096  11053.10967      Dengue virus 1 SF729VI06  Dengue virus   \n",
       "1112  11053.10981     Dengue virus 1 CNM664VI06  Dengue virus   \n",
       "1113  11053.10982     Dengue virus 1 CNM399VI07  Dengue virus   \n",
       "1157  11053.11020  Dengue virus 1 AO/DB136/2013  Dengue virus   \n",
       "1160  11053.11023  Dengue virus 1 AO/DB135/2013  Dengue virus   \n",
       "\n",
       "           Completion Date GenBank Accession  Size collection_date  \\\n",
       "1096  2016-07-24T00:00:00Z          FJ502874   216            2005   \n",
       "1112  2016-07-24T00:00:00Z          FJ502864   216            2006   \n",
       "1113  2016-07-24T00:00:00Z          FJ502863   216            2007   \n",
       "1157  2014-11-23T00:00:00Z          KM277614  1485            2013   \n",
       "1160  2014-11-23T00:00:00Z          KM277613  1485            2013   \n",
       "\n",
       "         country Geographic Group     Host Name Host Common Name  \n",
       "1096  Madagascar           Africa  Homo sapiens            Human  \n",
       "1112    Cameroon           Africa  Homo sapiens            Human  \n",
       "1113     Nigeria           Africa  Homo sapiens            Human  \n",
       "1157      Angola           Africa  Homo sapiens            Human  \n",
       "1160      Angola           Africa  Homo sapiens            Human  "
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_africa.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GenBank Accession</th>\n",
       "      <th>length</th>\n",
       "      <th>category_1</th>\n",
       "      <th>category_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>FJ502874</td>\n",
       "      <td>216</td>\n",
       "      <td>dengue virus type 1</td>\n",
       "      <td>DENV-1 Genotype IV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FJ502864</td>\n",
       "      <td>216</td>\n",
       "      <td>dengue virus type 1</td>\n",
       "      <td>Could not assign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FJ502863</td>\n",
       "      <td>216</td>\n",
       "      <td>dengue virus type 1</td>\n",
       "      <td>DENV-1 Genotype V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>KM277614</td>\n",
       "      <td>1485</td>\n",
       "      <td>dengue virus type 1</td>\n",
       "      <td>Could not assign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>KM277613</td>\n",
       "      <td>1485</td>\n",
       "      <td>dengue virus type 1</td>\n",
       "      <td>Could not assign</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  GenBank Accession  length           category_1          category_2\n",
       "0          FJ502874     216  dengue virus type 1  DENV-1 Genotype IV\n",
       "1          FJ502864     216  dengue virus type 1    Could not assign\n",
       "2          FJ502863     216  dengue virus type 1   DENV-1 Genotype V\n",
       "3          KM277614    1485  dengue virus type 1    Could not assign\n",
       "4          KM277613    1485  dengue virus type 1    Could not assign"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GD_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "final_data = pd.merge(df_africa, GD_data, on = \"GenBank Accession\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Genome ID</th>\n",
       "      <th>Genome Name</th>\n",
       "      <th>Species</th>\n",
       "      <th>Completion Date</th>\n",
       "      <th>GenBank Accession</th>\n",
       "      <th>Size</th>\n",
       "      <th>collection_date</th>\n",
       "      <th>country</th>\n",
       "      <th>Geographic Group</th>\n",
       "      <th>Host Name</th>\n",
       "      <th>Host Common Name</th>\n",
       "      <th>length</th>\n",
       "      <th>category_1</th>\n",
       "      <th>category_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>11053.10967</td>\n",
       "      <td>Dengue virus 1 SF729VI06</td>\n",
       "      <td>Dengue virus</td>\n",
       "      <td>2016-07-24T00:00:00Z</td>\n",
       "      <td>FJ502874</td>\n",
       "      <td>216</td>\n",
       "      <td>2005</td>\n",
       "      <td>Madagascar</td>\n",
       "      <td>Africa</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Human</td>\n",
       "      <td>216</td>\n",
       "      <td>dengue virus type 1</td>\n",
       "      <td>DENV-1 Genotype IV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>11053.10981</td>\n",
       "      <td>Dengue virus 1 CNM664VI06</td>\n",
       "      <td>Dengue virus</td>\n",
       "      <td>2016-07-24T00:00:00Z</td>\n",
       "      <td>FJ502864</td>\n",
       "      <td>216</td>\n",
       "      <td>2006</td>\n",
       "      <td>Cameroon</td>\n",
       "      <td>Africa</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Human</td>\n",
       "      <td>216</td>\n",
       "      <td>dengue virus type 1</td>\n",
       "      <td>Could not assign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>11053.10982</td>\n",
       "      <td>Dengue virus 1 CNM399VI07</td>\n",
       "      <td>Dengue virus</td>\n",
       "      <td>2016-07-24T00:00:00Z</td>\n",
       "      <td>FJ502863</td>\n",
       "      <td>216</td>\n",
       "      <td>2007</td>\n",
       "      <td>Nigeria</td>\n",
       "      <td>Africa</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Human</td>\n",
       "      <td>216</td>\n",
       "      <td>dengue virus type 1</td>\n",
       "      <td>DENV-1 Genotype V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11053.11020</td>\n",
       "      <td>Dengue virus 1 AO/DB136/2013</td>\n",
       "      <td>Dengue virus</td>\n",
       "      <td>2014-11-23T00:00:00Z</td>\n",
       "      <td>KM277614</td>\n",
       "      <td>1485</td>\n",
       "      <td>2013</td>\n",
       "      <td>Angola</td>\n",
       "      <td>Africa</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Human</td>\n",
       "      <td>1485</td>\n",
       "      <td>dengue virus type 1</td>\n",
       "      <td>Could not assign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11053.11023</td>\n",
       "      <td>Dengue virus 1 AO/DB135/2013</td>\n",
       "      <td>Dengue virus</td>\n",
       "      <td>2014-11-23T00:00:00Z</td>\n",
       "      <td>KM277613</td>\n",
       "      <td>1485</td>\n",
       "      <td>2013</td>\n",
       "      <td>Angola</td>\n",
       "      <td>Africa</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Human</td>\n",
       "      <td>1485</td>\n",
       "      <td>dengue virus type 1</td>\n",
       "      <td>Could not assign</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Genome ID                   Genome Name       Species  \\\n",
       "0  11053.10967      Dengue virus 1 SF729VI06  Dengue virus   \n",
       "1  11053.10981     Dengue virus 1 CNM664VI06  Dengue virus   \n",
       "2  11053.10982     Dengue virus 1 CNM399VI07  Dengue virus   \n",
       "3  11053.11020  Dengue virus 1 AO/DB136/2013  Dengue virus   \n",
       "4  11053.11023  Dengue virus 1 AO/DB135/2013  Dengue virus   \n",
       "\n",
       "        Completion Date GenBank Accession  Size collection_date     country  \\\n",
       "0  2016-07-24T00:00:00Z          FJ502874   216            2005  Madagascar   \n",
       "1  2016-07-24T00:00:00Z          FJ502864   216            2006    Cameroon   \n",
       "2  2016-07-24T00:00:00Z          FJ502863   216            2007     Nigeria   \n",
       "3  2014-11-23T00:00:00Z          KM277614  1485            2013      Angola   \n",
       "4  2014-11-23T00:00:00Z          KM277613  1485            2013      Angola   \n",
       "\n",
       "  Geographic Group     Host Name Host Common Name  length  \\\n",
       "0           Africa  Homo sapiens            Human     216   \n",
       "1           Africa  Homo sapiens            Human     216   \n",
       "2           Africa  Homo sapiens            Human     216   \n",
       "3           Africa  Homo sapiens            Human    1485   \n",
       "4           Africa  Homo sapiens            Human    1485   \n",
       "\n",
       "            category_1          category_2  \n",
       "0  dengue virus type 1  DENV-1 Genotype IV  \n",
       "1  dengue virus type 1    Could not assign  \n",
       "2  dengue virus type 1   DENV-1 Genotype V  \n",
       "3  dengue virus type 1    Could not assign  \n",
       "4  dengue virus type 1    Could not assign  "
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Genome ID</th>\n",
       "      <th>Genome Name</th>\n",
       "      <th>Species</th>\n",
       "      <th>Completion Date</th>\n",
       "      <th>GenBank Accession</th>\n",
       "      <th>Size</th>\n",
       "      <th>collection_date</th>\n",
       "      <th>country</th>\n",
       "      <th>Geographic Group</th>\n",
       "      <th>Host Name</th>\n",
       "      <th>Host Common Name</th>\n",
       "      <th>length</th>\n",
       "      <th>category_1</th>\n",
       "      <th>category_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>11053.10967</td>\n",
       "      <td>Dengue virus 1 SF729VI06</td>\n",
       "      <td>Dengue virus</td>\n",
       "      <td>2016-07-24T00:00:00Z</td>\n",
       "      <td>FJ502874</td>\n",
       "      <td>216</td>\n",
       "      <td>2005</td>\n",
       "      <td>Madagascar</td>\n",
       "      <td>Africa</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Human</td>\n",
       "      <td>216</td>\n",
       "      <td>dengue virus type 1</td>\n",
       "      <td>DENV-1 Genotype IV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>11053.10981</td>\n",
       "      <td>Dengue virus 1 CNM664VI06</td>\n",
       "      <td>Dengue virus</td>\n",
       "      <td>2016-07-24T00:00:00Z</td>\n",
       "      <td>FJ502864</td>\n",
       "      <td>216</td>\n",
       "      <td>2006</td>\n",
       "      <td>Cameroon</td>\n",
       "      <td>Africa</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Human</td>\n",
       "      <td>216</td>\n",
       "      <td>dengue virus type 1</td>\n",
       "      <td>Could not assign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>11053.10982</td>\n",
       "      <td>Dengue virus 1 CNM399VI07</td>\n",
       "      <td>Dengue virus</td>\n",
       "      <td>2016-07-24T00:00:00Z</td>\n",
       "      <td>FJ502863</td>\n",
       "      <td>216</td>\n",
       "      <td>2007</td>\n",
       "      <td>Nigeria</td>\n",
       "      <td>Africa</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Human</td>\n",
       "      <td>216</td>\n",
       "      <td>dengue virus type 1</td>\n",
       "      <td>DENV-1 Genotype V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11053.11020</td>\n",
       "      <td>Dengue virus 1 AO/DB136/2013</td>\n",
       "      <td>Dengue virus</td>\n",
       "      <td>2014-11-23T00:00:00Z</td>\n",
       "      <td>KM277614</td>\n",
       "      <td>1485</td>\n",
       "      <td>2013</td>\n",
       "      <td>Angola</td>\n",
       "      <td>Africa</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Human</td>\n",
       "      <td>1485</td>\n",
       "      <td>dengue virus type 1</td>\n",
       "      <td>Could not assign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11053.11023</td>\n",
       "      <td>Dengue virus 1 AO/DB135/2013</td>\n",
       "      <td>Dengue virus</td>\n",
       "      <td>2014-11-23T00:00:00Z</td>\n",
       "      <td>KM277613</td>\n",
       "      <td>1485</td>\n",
       "      <td>2013</td>\n",
       "      <td>Angola</td>\n",
       "      <td>Africa</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Human</td>\n",
       "      <td>1485</td>\n",
       "      <td>dengue virus type 1</td>\n",
       "      <td>Could not assign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>868</th>\n",
       "      <td>12637.37810</td>\n",
       "      <td>Dengue virus DENV-2/TAN/IST/2014</td>\n",
       "      <td>Dengue virus</td>\n",
       "      <td>2016-12-31T00:00:00Z</td>\n",
       "      <td>KM892495</td>\n",
       "      <td>500</td>\n",
       "      <td>2014-02</td>\n",
       "      <td>Tanzania</td>\n",
       "      <td>Africa</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Human</td>\n",
       "      <td>500</td>\n",
       "      <td>dengue virus type 2</td>\n",
       "      <td>DENV-2 Genotype II - Cosmopolitan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>869</th>\n",
       "      <td>12637.62300</td>\n",
       "      <td>Dengue virus DENV-2/TAN/KCMC/2014</td>\n",
       "      <td>Dengue virus</td>\n",
       "      <td>2018-03-01T00:00:00Z</td>\n",
       "      <td>KU886233</td>\n",
       "      <td>500</td>\n",
       "      <td>2014-06</td>\n",
       "      <td>Tanzania</td>\n",
       "      <td>Africa</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Human</td>\n",
       "      <td>500</td>\n",
       "      <td>dengue virus type 2</td>\n",
       "      <td>DENV-2 Genotype II - Cosmopolitan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>870</th>\n",
       "      <td>12637.72700</td>\n",
       "      <td>Dengue virus</td>\n",
       "      <td>Dengue virus</td>\n",
       "      <td>2016-07-26T00:00:00Z</td>\n",
       "      <td>DQ016638</td>\n",
       "      <td>306</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Kenya</td>\n",
       "      <td>Africa</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Human</td>\n",
       "      <td>306</td>\n",
       "      <td>dengue virus type 1</td>\n",
       "      <td>DENV-1 Genotype I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>871</th>\n",
       "      <td>12637.73300</td>\n",
       "      <td>Dengue virus</td>\n",
       "      <td>Dengue virus</td>\n",
       "      <td>2016-07-26T00:00:00Z</td>\n",
       "      <td>DQ016641</td>\n",
       "      <td>306</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Kenya</td>\n",
       "      <td>Africa</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Human</td>\n",
       "      <td>306</td>\n",
       "      <td>dengue virus type 1</td>\n",
       "      <td>DENV-1 Genotype I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>872</th>\n",
       "      <td>12637.73700</td>\n",
       "      <td>Dengue virus</td>\n",
       "      <td>Dengue virus</td>\n",
       "      <td>2016-07-26T00:00:00Z</td>\n",
       "      <td>DQ016643</td>\n",
       "      <td>306</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sudan</td>\n",
       "      <td>Africa</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Human</td>\n",
       "      <td>306</td>\n",
       "      <td>dengue virus type 2</td>\n",
       "      <td>DENV-2 Genotype V - Asian I</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>873 rows Ã— 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Genome ID                        Genome Name       Species  \\\n",
       "0    11053.10967           Dengue virus 1 SF729VI06  Dengue virus   \n",
       "1    11053.10981          Dengue virus 1 CNM664VI06  Dengue virus   \n",
       "2    11053.10982          Dengue virus 1 CNM399VI07  Dengue virus   \n",
       "3    11053.11020       Dengue virus 1 AO/DB136/2013  Dengue virus   \n",
       "4    11053.11023       Dengue virus 1 AO/DB135/2013  Dengue virus   \n",
       "..           ...                                ...           ...   \n",
       "868  12637.37810   Dengue virus DENV-2/TAN/IST/2014  Dengue virus   \n",
       "869  12637.62300  Dengue virus DENV-2/TAN/KCMC/2014  Dengue virus   \n",
       "870  12637.72700                       Dengue virus  Dengue virus   \n",
       "871  12637.73300                       Dengue virus  Dengue virus   \n",
       "872  12637.73700                       Dengue virus  Dengue virus   \n",
       "\n",
       "          Completion Date GenBank Accession  Size collection_date     country  \\\n",
       "0    2016-07-24T00:00:00Z          FJ502874   216            2005  Madagascar   \n",
       "1    2016-07-24T00:00:00Z          FJ502864   216            2006    Cameroon   \n",
       "2    2016-07-24T00:00:00Z          FJ502863   216            2007     Nigeria   \n",
       "3    2014-11-23T00:00:00Z          KM277614  1485            2013      Angola   \n",
       "4    2014-11-23T00:00:00Z          KM277613  1485            2013      Angola   \n",
       "..                    ...               ...   ...             ...         ...   \n",
       "868  2016-12-31T00:00:00Z          KM892495   500         2014-02    Tanzania   \n",
       "869  2018-03-01T00:00:00Z          KU886233   500         2014-06    Tanzania   \n",
       "870  2016-07-26T00:00:00Z          DQ016638   306             NaN       Kenya   \n",
       "871  2016-07-26T00:00:00Z          DQ016641   306             NaN       Kenya   \n",
       "872  2016-07-26T00:00:00Z          DQ016643   306             NaN       Sudan   \n",
       "\n",
       "    Geographic Group     Host Name Host Common Name  length  \\\n",
       "0             Africa  Homo sapiens            Human     216   \n",
       "1             Africa  Homo sapiens            Human     216   \n",
       "2             Africa  Homo sapiens            Human     216   \n",
       "3             Africa  Homo sapiens            Human    1485   \n",
       "4             Africa  Homo sapiens            Human    1485   \n",
       "..               ...           ...              ...     ...   \n",
       "868           Africa  Homo sapiens            Human     500   \n",
       "869           Africa  Homo sapiens            Human     500   \n",
       "870           Africa  Homo sapiens            Human     306   \n",
       "871           Africa  Homo sapiens            Human     306   \n",
       "872           Africa  Homo sapiens            Human     306   \n",
       "\n",
       "              category_1                         category_2  \n",
       "0    dengue virus type 1                 DENV-1 Genotype IV  \n",
       "1    dengue virus type 1                   Could not assign  \n",
       "2    dengue virus type 1                  DENV-1 Genotype V  \n",
       "3    dengue virus type 1                   Could not assign  \n",
       "4    dengue virus type 1                   Could not assign  \n",
       "..                   ...                                ...  \n",
       "868  dengue virus type 2  DENV-2 Genotype II - Cosmopolitan  \n",
       "869  dengue virus type 2  DENV-2 Genotype II - Cosmopolitan  \n",
       "870  dengue virus type 1                  DENV-1 Genotype I  \n",
       "871  dengue virus type 1                  DENV-1 Genotype I  \n",
       "872  dengue virus type 2        DENV-2 Genotype V - Asian I  \n",
       "\n",
       "[873 rows x 14 columns]"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# dropping duplicates\n",
    "\n",
    "final_data.drop_duplicates([\"GenBank Accession\"], keep = \"first\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False    873\n",
       "Name: collection_date, dtype: int64"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# checking if the data has -N/A- in collection_date col\n",
    "\n",
    "final_data[\"collection_date\"].isin([\"NaN\"]).value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "# if true, replacing -N/A- with null values \n",
    "\n",
    "# final_df[\"collection_date\"] = final_df[\"collection_date\"].replace(\"NaN\", None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "final_data[\"collection_date\"] = pd.to_datetime(final_data[\"collection_date\"], yearfirst = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0   2005-01-01\n",
       "1   2006-01-01\n",
       "2   2007-01-01\n",
       "3   2013-01-01\n",
       "4   2013-01-01\n",
       "Name: collection_date, dtype: datetime64[ns]"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_data[\"collection_date\"].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Changing the Serotypes and Genotypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# checking if there are NaNs\n",
    "\n",
    "final_data[\"category_1\"].hasnans"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "# if True run...\n",
    "\n",
    "# final_df[\"lineages\"] = final_df[\"lineages\"].fillna(\"Unassigned\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dengue virus type 1    458\n",
       "dengue virus type 2    256\n",
       "dengue virus type 3    136\n",
       "dengue virus type 4     15\n",
       "Could not assign         8\n",
       "Name: category_1, dtype: int64"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_data[\"category_1\"].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For this tutorial we are only interested on Dengue virus type 1. So, let's filter the dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "dengue_1 = final_data[final_data[\"category_1\"] == \"dengue virus type 1\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Related to but not part of DENV-1 Genotype III    362\n",
       "Could not assign                                   45\n",
       "DENV-1 Genotype I                                  26\n",
       "DENV-1 Genotype IV                                 11\n",
       "DENV-1 Genotype V                                   6\n",
       "DENV-1 Genotype II                                  5\n",
       "DENV-1 Genotype III                                 2\n",
       "Related to but not part of DENV-1 Genotype V        1\n",
       "Name: category_2, dtype: int64"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dengue_1[\"category_2\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "# defining function to change genotype names\n",
    "# row number needs to be the col = variants\n",
    "\n",
    "def standardize_genotype(row):\n",
    "    genotype = row[\"category_2\"]\n",
    "    \n",
    "    if genotype == \"Related to but not part of DENV-1 Genotype III\":\n",
    "        return \"Unassigned\"\n",
    "    elif genotype == \"DENV-1 Genotype I\":\n",
    "        return \"Genotype I\"\n",
    "    elif genotype == \"DENV-1 Genotype II\":\n",
    "        return \"Genotype II\"\n",
    "    elif genotype == \"DENV-1 Genotype III\":\n",
    "        return \"Genotype III\"\n",
    "    elif genotype == \"DENV-1 Genotype IV\":\n",
    "        return \"Genotype IV\"\n",
    "    elif genotype == \"DENV-1 Genotype V\":\n",
    "        return \"Genotype V\"\n",
    "    elif genotype == \"Could not assign\":\n",
    "        return \"Unassigned\"\n",
    "    else:\n",
    "        return genotype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_133/2468620654.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  dengue_1[\"category_2\"] = dengue_1.apply(standardize_genotype, axis = \"columns\")\n"
     ]
    }
   ],
   "source": [
    "dengue_1[\"category_2\"] = dengue_1.apply(standardize_genotype, axis = \"columns\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Genome ID</th>\n",
       "      <th>Genome Name</th>\n",
       "      <th>Species</th>\n",
       "      <th>Completion Date</th>\n",
       "      <th>GenBank Accession</th>\n",
       "      <th>Size</th>\n",
       "      <th>collection_date</th>\n",
       "      <th>country</th>\n",
       "      <th>Geographic Group</th>\n",
       "      <th>Host Name</th>\n",
       "      <th>Host Common Name</th>\n",
       "      <th>length</th>\n",
       "      <th>category_1</th>\n",
       "      <th>category_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>11053.10967</td>\n",
       "      <td>Dengue virus 1 SF729VI06</td>\n",
       "      <td>Dengue virus</td>\n",
       "      <td>2016-07-24T00:00:00Z</td>\n",
       "      <td>FJ502874</td>\n",
       "      <td>216</td>\n",
       "      <td>2005-01-01</td>\n",
       "      <td>Madagascar</td>\n",
       "      <td>Africa</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Human</td>\n",
       "      <td>216</td>\n",
       "      <td>dengue virus type 1</td>\n",
       "      <td>Genotype IV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>11053.10981</td>\n",
       "      <td>Dengue virus 1 CNM664VI06</td>\n",
       "      <td>Dengue virus</td>\n",
       "      <td>2016-07-24T00:00:00Z</td>\n",
       "      <td>FJ502864</td>\n",
       "      <td>216</td>\n",
       "      <td>2006-01-01</td>\n",
       "      <td>Cameroon</td>\n",
       "      <td>Africa</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Human</td>\n",
       "      <td>216</td>\n",
       "      <td>dengue virus type 1</td>\n",
       "      <td>Unassigned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>11053.10982</td>\n",
       "      <td>Dengue virus 1 CNM399VI07</td>\n",
       "      <td>Dengue virus</td>\n",
       "      <td>2016-07-24T00:00:00Z</td>\n",
       "      <td>FJ502863</td>\n",
       "      <td>216</td>\n",
       "      <td>2007-01-01</td>\n",
       "      <td>Nigeria</td>\n",
       "      <td>Africa</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Human</td>\n",
       "      <td>216</td>\n",
       "      <td>dengue virus type 1</td>\n",
       "      <td>Genotype V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11053.11020</td>\n",
       "      <td>Dengue virus 1 AO/DB136/2013</td>\n",
       "      <td>Dengue virus</td>\n",
       "      <td>2014-11-23T00:00:00Z</td>\n",
       "      <td>KM277614</td>\n",
       "      <td>1485</td>\n",
       "      <td>2013-01-01</td>\n",
       "      <td>Angola</td>\n",
       "      <td>Africa</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Human</td>\n",
       "      <td>1485</td>\n",
       "      <td>dengue virus type 1</td>\n",
       "      <td>Unassigned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11053.11023</td>\n",
       "      <td>Dengue virus 1 AO/DB135/2013</td>\n",
       "      <td>Dengue virus</td>\n",
       "      <td>2014-11-23T00:00:00Z</td>\n",
       "      <td>KM277613</td>\n",
       "      <td>1485</td>\n",
       "      <td>2013-01-01</td>\n",
       "      <td>Angola</td>\n",
       "      <td>Africa</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Human</td>\n",
       "      <td>1485</td>\n",
       "      <td>dengue virus type 1</td>\n",
       "      <td>Unassigned</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Genome ID                   Genome Name       Species  \\\n",
       "0  11053.10967      Dengue virus 1 SF729VI06  Dengue virus   \n",
       "1  11053.10981     Dengue virus 1 CNM664VI06  Dengue virus   \n",
       "2  11053.10982     Dengue virus 1 CNM399VI07  Dengue virus   \n",
       "3  11053.11020  Dengue virus 1 AO/DB136/2013  Dengue virus   \n",
       "4  11053.11023  Dengue virus 1 AO/DB135/2013  Dengue virus   \n",
       "\n",
       "        Completion Date GenBank Accession  Size collection_date     country  \\\n",
       "0  2016-07-24T00:00:00Z          FJ502874   216      2005-01-01  Madagascar   \n",
       "1  2016-07-24T00:00:00Z          FJ502864   216      2006-01-01    Cameroon   \n",
       "2  2016-07-24T00:00:00Z          FJ502863   216      2007-01-01     Nigeria   \n",
       "3  2014-11-23T00:00:00Z          KM277614  1485      2013-01-01      Angola   \n",
       "4  2014-11-23T00:00:00Z          KM277613  1485      2013-01-01      Angola   \n",
       "\n",
       "  Geographic Group     Host Name Host Common Name  length  \\\n",
       "0           Africa  Homo sapiens            Human     216   \n",
       "1           Africa  Homo sapiens            Human     216   \n",
       "2           Africa  Homo sapiens            Human     216   \n",
       "3           Africa  Homo sapiens            Human    1485   \n",
       "4           Africa  Homo sapiens            Human    1485   \n",
       "\n",
       "            category_1   category_2  \n",
       "0  dengue virus type 1  Genotype IV  \n",
       "1  dengue virus type 1   Unassigned  \n",
       "2  dengue virus type 1   Genotype V  \n",
       "3  dengue virus type 1   Unassigned  \n",
       "4  dengue virus type 1   Unassigned  "
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dengue_1.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_133/311299808.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  dengue_1[\"date\"] = dengue_1[\"collection_date\"]\n"
     ]
    }
   ],
   "source": [
    "# Setting date column name\n",
    "dengue_1[\"date\"] = dengue_1[\"collection_date\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "output_file_path = \"[PATH TO THE FILE]\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "dengue_1.to_csv(\"dengue_1.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<a href='genbank_accession.txt' target='_blank'>genbank_accession.txt</a><br>"
      ],
      "text/plain": [
       "/home/jupyter/VEME 2023 Pathogen Dashboards/edit/genbank_accession.txt"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(FileLink(dengue_1))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {
    "height": "102px",
    "width": "337px"
   },
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
